pdns/tcpiohandler.cc

```diff
@@ -1817,46 +1817,57 @@ bool setupDoTProtocolNegotiation(std::shared_ptr<TLSCtx>& ctx)
   return true;
 }
 
+bool setupDoHProtocolNegotiation(std::shared_ptr<TLSCtx>& ctx)
+{
+  if (ctx == nullptr) {
+    return false;
+  }
+  /* we want to set the ALPN to doh */
+  const std::vector<std::vector<uint8_t>> dohAlpns = {{'h', '2'}};
+  ctx->setALPNProtos(dohAlpns);
+  return true;
+}
+
 bool TLSFrontend::setupTLS()
 {
-#ifdef HAVE_DNS_OVER_TLS
+#if defined(HAVE_DNS_OVER_TLS) || defined(HAVE_DNS_OVER_HTTPS)
   std::shared_ptr<TLSCtx> newCtx{nullptr};
   /* get the "best" available provider */
-  if (!d_provider.empty()) {
 #ifdef HAVE_GNUTLS
-    if (d_provider == "gnutls") {
-      newCtx = std::make_shared<GnuTLSIOCtx>(*this);
-      setupDoTProtocolNegotiation(newCtx);
-      std::atomic_store_explicit(&d_ctx, newCtx, std::memory_order_release);
-      return true;
-    }
+  if (d_provider == "gnutls") {
+    newCtx = std::make_shared<GnuTLSIOCtx>(*this);
+  }
 #endif /* HAVE_GNUTLS */
 #ifdef HAVE_LIBSSL
-    if (d_provider == "openssl") {
-      newCtx = std::make_shared<OpenSSLTLSIOCtx>(*this);
-      setupDoTProtocolNegotiation(newCtx);
-      std::atomic_store_explicit(&d_ctx, newCtx, std::memory_order_release);
-      return true;
-    }
-#endif /* HAVE_LIBSSL */
+  if (d_provider == "openssl") {
+    newCtx = std::make_shared<OpenSSLTLSIOCtx>(*this);
   }
+#endif /* HAVE_LIBSSL */
+  if (!newCtx) {
 #ifdef HAVE_LIBSSL
-  newCtx = std::make_shared<OpenSSLTLSIOCtx>(*this);
+    newCtx = std::make_shared<OpenSSLTLSIOCtx>(*this);
 #else /* HAVE_LIBSSL */
 #ifdef HAVE_GNUTLS
-  newCtx = std::make_shared<GnuTLSIOCtx>(*this);
+    newCtx = std::make_shared<GnuTLSIOCtx>(*this);
 #endif /* HAVE_GNUTLS */
 #endif /* HAVE_LIBSSL */
```

Comment:
>  I don't remember seeing anything in the auth configure.ac and accompanying scripts that stops from enabling both libssl and gnuTLS. Also running dnsdist's configure script with --with-libssl and --with-gnutls shows that it's possible to enable both.

Correct, we do want to be able to enable both of them at compile at time, but also at runtime! This code is only used if the user did not explicitly select one in the configuration.

> I would then prefer the use of `#elif` over `#endif`+`#ifdef` and the nested `#else`+`#ifdef` since it makes it more clear which library we prioritize and the respective library calls are always in the same order (e.g. libssl-related code will always come before GnuTLS-related code).

It makes a lot of sense, yes. I need to check whether this code is actually expected to compile when neither libssl nor GnuTLS are available, because if it does the final `#else` should go, but the rest looks good at first glance. Thanks!

Reply:
Done now whenever possible.